### DISCUSSION TITLE
Allow custom headers in "upgrade" response

### DISCUSSION THREAD

**Discussion by not-implemented** [uNetworking/uWebSockets#1075] [Q&A] [answered]
There was no possibility so far, to set custom headers (e.g. for diagnostic purposes) in the HTTP "upgrade" response. With the cool new "upgrade" handler, it seems to be much easier now, to integrate such an interface (currently it's still impossible without patching, right?). Maybe as a parameter/callback of `res->upgrade()` or something similar?

**Comment by ghost** [comment]
https://unetworking.github.io/uWebSockets.js/generated/interfaces/httpresponse.html#writestatus

You can do it but it's not obvious

**Comment by not-implemented** [comment]
But I can't without reimplementing `HttpResponse::upgrade()`, right? Because I can't hook here in between `writeStatus()` and `internalEnd()` later:

https://github.com/uNetworking/uWebSockets/blob/master/src/HttpResponse.h#L195

**Comment by ghost** [comment]
writeStatus can only be called once, so second time will be ignored.

writeStatus, writeHeader, upgrade is the order you have to use

**Comment by not-implemented** [comment, marked as answer]
Ah I see ... ok ... so I just have to "duplicate" the `writeStatus("101 Switching Protocols")` line ... okay ... that's ... acceptable ;-)

Thanks!

**Comment by ghost** [comment]
Yes that is not very obvious but it works for now. Upgrade could take a callback like cork does, in which you writeHeader but that's for another time

**Comment by not-implemented** [comment]
Yes.
